feat: add `stats/base/ndarray/dmidrange`

DivyanshuVortex opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed
Resolves None.

---
## Description

This pull request:

feat : add stats/base/ndarray/dmidrange

## Related Issues

This pull request has the following related issues:

- None .

## Questions

No.

## Other

No.

## Checklist

- [x] Read, understood, and followed the [contributing guidelines][contributing].

### AI Assistance

- [x] Yes  
- [ ] No

How was AI assistance used?

- [ ] Code generation  
- [ ] Test/benchmark generation  
- [ ] Documentation  
- [x] Research and understanding

#### Disclosure

I consulted ChatGPT for clarification on documentation format and contribution structure. All implementation logic, tests, and documentation were written manually.

---

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

DivyanshuVortex committed:
docs(ndarray-dmidrange): update readme, examples, and doctest output for midrange implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

DivyanshuVortex committed:
chores

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/README.md:
```diff
@@ -0,0 +1,140 @@
+
+# dmidrange
+
+> Compute the [midrange][midrange] of a one-dimensional double-precision floating-point ndarray, ignoring `NaN` values .
```
This isn't correct. `dmidrange` does not ignore `NaN` values. That package would be `dnanmidrange`, for which the underlying strided array package has not yet been implemented.

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/README.md:
```diff
@@ -0,0 +1,140 @@
+var dmidrange = require( '@stdlib/stats/base/ndarray/dmidrange' );
+
+function rand() {
+    if ( bernoulli( 0.8 ) === 1 ) {
```
1. We should not be generating NaN elements for this package as the result will always be NaN.
2. Why strict equality here? In this case, given `0.8`, it is very likely that `rand` will return `NaN`. There is a reason why we use `<` elsewhere, as that means `NaN` values are only 20% likely to happen, not 80% likely to happen as you have it now.

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/benchmark/benchmark.js:
```diff
@@ -0,0 +1,103 @@
+	var xbuf;
+	var x;
+
+	xbuf = filledarrayBy( len, 'float64', rand );
```
You don't need to use `filledarrayBy` and `new ndarray` here. You can use `random/uniform` directly.

kgryte committed:
chore: remove incorrect file

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/docs/repl.txt:
```diff
@@ -0,0 +1,32 @@
+
+{{alias}}( arrays )
+    Computes the midrange of a one-dimensional double-precision floating-
+    point ndarray, ignoring `NaN` values.
```
Incorrect description.

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/docs/types/index.d.ts:
```diff
@@ -0,0 +1,46 @@
+import { float64ndarray } from '@stdlib/types/ndarray';
+
+/**
+* Computes the midrange value of a one-dimensional double-precision floating-point ndarray, ignoring `NaN` values.
```
Description.

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/examples/index.js:
```diff
@@ -0,0 +1,40 @@
+var ndarray2array = require( '@stdlib/ndarray/to-array' );
+var dmidrange = require( './../lib' );
+
+function rand() {
```
Not desired. Consider using `random/uniform` directly.

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/lib/index.js:
```diff
@@ -0,0 +1,46 @@
+'use strict';
+
+/**
+* Compute the midrange of a one-dimensional double-precision floating-point ndarray If a `NaN` value is encountered during iteration, the.
```
Incorrect description. Use consistent descriptions.

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/lib/main.js:
```diff
@@ -0,0 +1,56 @@
+// MAIN //
+
+/**
+* Computes the arithmetic mean of a one-dimensional double-precision floating-point ndarray, ignoring `NaN` values and using ordinary recursive summation.
```
Incorrect description.

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/lib/main.js:
```diff
@@ -0,0 +1,56 @@
+* Computes the arithmetic mean of a one-dimensional double-precision floating-point ndarray, ignoring `NaN` values and using ordinary recursive summation.
+*
+* @param {ArrayLikeObject<Object>} arrays - array-like object containing an input ndarray
+* @returns {number} arithmetic mean
```
Incorrect.

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/package.json:
```diff
@@ -0,0 +1,67 @@
+{
+  "name": "@stdlib/stats/base/ndarray/dmidrange",
+  "version": "0.0.0",
+  "description": "Compute the arithmetic mean of a one-dimensional double-precision floating-point ndarray, ignoring `NaN` values and using ordinary recursive summation.",
```
Description.

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/test/test.js:
```diff
@@ -0,0 +1,184 @@
+	t.end();
+});
+
+tape( 'the function returns NaN when encountering NaN values', function test( t ) {
```
This test should have been a red flag that most of your descriptions were incorrect.

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/test/test.js:
```diff
@@ -0,0 +1,184 @@
+
+	x = new Float64Array( [ 1.0, -2.0, NaN, 2.0 ] );
+
+	v = dmidrange( [ vector( x, 4, 1, 0 ) ] );
```
You are lacking tests which actually demonstrate that the function computes the midrange.

kgryte replied:
@DivyanshuVortex These tests still need updating. See the `stats/strided/dmidrange` package tests for inspiration.

kgryte replied:
This still needs updating.

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/test/test.js:
```diff
@@ -0,0 +1,184 @@
+	]);
+
+	v = dmidrange( [ vector( x, 5, 2, 0 ) ] );
+	t.strictEqual( isnan( v ), true, 'returns NaN due to NaN in scanned path' );
```
This shouldn't be the case. You needed to copy-paste a different package.

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/test/test.js:
```diff
@@ -0,0 +1,184 @@
+		3.0,
+		4.0,
+		2.0,
+		NaN
```
This test needs updating.

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/test/test.js:
```diff
@@ -0,0 +1,184 @@
+		2.0,  // 2
+		3.0,
+		4.0,  // 3
+		NaN,
```
This test needs updating.

kgryte reviewed (changes requested):
Left an initial round of comments. This PR needs significant clean-up before it can move forward.

DivyanshuVortex committed:
descrp

DivyanshuVortex committed:
descrp

DivyanshuVortex committed:
descrp

DivyanshuVortex committed:
descrp

DivyanshuVortex committed:
descrp

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/benchmark/benchmark.js:
```diff
@@ -0,0 +1,99 @@
+
+		b.tic();
+		for ( i = 0; i < b.iterations; i++ ) {
+			v = dmidrange( [ x ] );
```
You are missing an `isnan` assertion check within the loop. See other packages.

kgryte committed:
docs: fix description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/docs/types/index.d.ts:
```diff
@@ -0,0 +1,46 @@
+import { float64ndarray } from '@stdlib/types/ndarray';
+
+/**
+* Compute the midrange value of a one-dimensional double-precision floating-point ndarray.
```
```suggestion
* Compute the mid-range value of a one-dimensional double-precision floating-point ndarray.
```

Use descriptions and spelling which are consistent with other packages (e.g., `stats/strided/dmidrange`).

kgryte replied:
This applies here and everywhere else in this PR.

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/README.md:
```diff
@@ -0,0 +1,120 @@
+var ndarray2array = require( '@stdlib/ndarray/to-array' );
+var dmidrange = require( '@stdlib/stats/base/ndarray/dmidrange' );
+
+var xbuf = uniform( 10, -50.0, 50.0 );
```
```suggestion
var xbuf = uniform( 10, -50.0, 50.0, {
    'dtype': 'float64'
});
```

kgryte committed:
docs: update example

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/examples/index.js:
```diff
@@ -0,0 +1,31 @@
+var ndarray2array = require( '@stdlib/ndarray/to-array' );
+var dmidrange = require( './../lib' );
+
+var xbuf = uniform( 10, -50.0, 50.0 );
```
```suggestion
var xbuf = uniform( 10, -50.0, 50.0, {
	'dtype': 'float64'
});
```

kgryte committed:
docs: explicit set dtype

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/lib/index.js:
```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+/**
+* Compute the midrange of a one-dimensional double-precision floating-point ndarray.
```
```suggestion
* Compute the mid-range of a one-dimensional double-precision floating-point ndarray.
```

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/lib/index.js:
```diff
@@ -0,0 +1,45 @@
+* var ndarray = require( '@stdlib/ndarray/base/ctor' );
+* var dmidrange = require( '@stdlib/stats/base/ndarray/dmidrange' );
+*
+* var xbuf = new Float64Array( [ 1.0, -2.0, NaN, 2.0 ] );
```
This is incorrect.

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/lib/index.js:
```diff
@@ -0,0 +1,45 @@
+* var x = new ndarray( 'float64', xbuf, [ 4 ], [ 1 ], 0, 'row-major' );
+*
+* var v = dmidrange( [ x ] );
+* // returns NaN
```
This is incorrect.

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/lib/main.js:
```diff
@@ -0,0 +1,56 @@
+// MAIN //
+
+/**
+* Compute the midrange of a one-dimensional double-precision floating-point ndarray.
```
```suggestion
* Compute the mid-range of a one-dimensional double-precision floating-point ndarray.
```

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/lib/main.js:
```diff
@@ -0,0 +1,56 @@
+* Compute the midrange of a one-dimensional double-precision floating-point ndarray.
+*
+* @param {ArrayLikeObject<Object>} arrays - array-like object containing an input ndarray
+* @returns {number} midrange value
```
```suggestion
* @returns {number} mid-range
```

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/package.json:
```diff
@@ -0,0 +1,67 @@
+{
+  "name": "@stdlib/stats/base/ndarray/dmidrange",
+  "version": "0.0.0",
+  "description": "Compute the midrange of a one-dimensional double-precision floating-point ndarray.",
```
```suggestion
  "description": "Compute the mid-range of a one-dimensional double-precision floating-point ndarray.",
```

kgryte reviewed (changes requested):
Left another round of comments.

DivyanshuVortex committed:
feat(ndarray,dmidrange): fix issues in fromat

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

DivyanshuVortex commented:
hi @kgryte ,
CI EditorConfig check is failing due to GitHub API rate limit while downloading `editorconfig-checker`. Local lint passes. Can someone please re-run or update the workflow?

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/docs/types/index.d.ts:
```diff
@@ -0,0 +1,46 @@
+import { float64ndarray } from '@stdlib/types/ndarray';
+
+/**
+* Compute the mid-range value of a one-dimensional double-precision floating-point ndarray.
```
```suggestion
* Computes the mid-range value of a one-dimensional double-precision floating-point ndarray.
```

kgryte committed:
docs: fix desc

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/docs/types/index.d.ts:
```diff
@@ -0,0 +1,46 @@
+import { float64ndarray } from '@stdlib/types/ndarray';
+
+/**
+* Computes the mid-range value of a one-dimensional double-precision floating-point ndarray.
```
```suggestion
* Computes the mid-range of a one-dimensional double-precision floating-point ndarray.
```

kgryte committed:
docs: fix desc

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/docs/types/index.d.ts:
```diff
@@ -0,0 +1,46 @@
+* Computes the mid-range of a one-dimensional double-precision floating-point ndarray.
+*
+* @param arrays - array-like object containing an input ndarray
+* @returns mid-range value
```
```suggestion
* @returns mid-range
```

kgryte committed:
docs: update desc

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/lib/main.js:
```diff
@@ -0,0 +1,56 @@
+* Compute the mid-range of a one-dimensional double-precision floating-point ndarray.
+*
+* @param {ArrayLikeObject<Object>} arrays - array-like object containing an input ndarray
+* @returns {number} mid-range value
```
```suggestion
* @returns {number} mid-range
```

kgryte committed:
docs: update desc

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/lib/main.js:
```diff
@@ -0,0 +1,56 @@
+// MAIN //
+
+/**
+* Compute the mid-range of a one-dimensional double-precision floating-point ndarray.
```
```suggestion
* Computes the mid-range of a one-dimensional double-precision floating-point ndarray.
```

kgryte committed:
docs: fix desc

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (changes requested):
Apart from tests, this PR is shaping up and close to ready for merge.

DivyanshuVortex committed:
docs: improve tests for dmidrange

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (approved):
LGTM
